source/opt/loop_dependence_helpers.cpp

```diff
@@ -0,0 +1,456 @@
+    // If distance - bounds > 0 we prove the distance is outwith the loop
+    // bounds.
+    if (distance_minus_bounds->FoldToSingleValue() > 0) {
+      PrintDebug(
+          "IsProvablyOutwithLoopBounds found distance escaped the loop "
+          "bounds.");
+      return true;
+    }
+  }
+
+  return false;
+}
+
+const ir::Loop* LoopDependenceAnalysis::GetLoopForSubscriptPair(
+    std::pair<SENode*, SENode*>* subscript_pair) {
+  // Collect all the SERecurrentNodes.
+  std::vector<SERecurrentNode*> source_nodes =
+      std::get<0>(*subscript_pair)->CollectRecurrentNodes();
+  std::vector<SERecurrentNode*> destination_nodes =
+      std::get<1>(*subscript_pair)->CollectRecurrentNodes();
+
+  // Collect all the loops stored by the SERecurrentNodes.
+  std::unordered_set<const ir::Loop*> loops{};
+  for (auto source_nodes_it = source_nodes.begin();
+       source_nodes_it != source_nodes.end(); ++source_nodes_it) {
+    loops.insert((*source_nodes_it)->GetLoop());
+  }
+  for (auto destination_nodes_it = destination_nodes.begin();
+       destination_nodes_it != destination_nodes.end();
+       ++destination_nodes_it) {
+    loops.insert((*destination_nodes_it)->GetLoop());
+  }
+
+  // If we didn't find 1 loop |subscript_pair| is a subscript over multiple or 0
+  // loops. We don't handle this so return nullptr.
+  if (loops.size() != 1) {
+    PrintDebug("GetLoopForSubscriptPair found loops.size() != 1.");
+    return nullptr;
+  }
+  return *loops.begin();
+}
+
+DistanceEntry* LoopDependenceAnalysis::GetDistanceEntryForSubscriptPair(
+    std::pair<SENode*, SENode*>* subscript_pair,
+    DistanceVector* distance_vector) {
+  const ir::Loop* loop = GetLoopForSubscriptPair(subscript_pair);
+  if (!loop) {
+    return nullptr;
+  }
+
+  DistanceEntry* distance_entry = nullptr;
+  for (size_t loop_index = 0; loop_index < loops_.size(); ++loop_index) {
+    if (loop == loops_[loop_index]) {
+      distance_entry = &distance_vector->entries[loop_index];
+      break;
+    }
+  }
+
+  return distance_entry;
+}
+
+SENode* LoopDependenceAnalysis::GetTripCount(const ir::Loop* loop) {
+  ir::BasicBlock* condition_block = loop->FindConditionBlock();
+  if (!condition_block) {
+    return nullptr;
+  }
+  ir::Instruction* induction_instr =
+      loop->FindConditionVariable(condition_block);
+  if (!induction_instr) {
+    return nullptr;
+  }
+  ir::Instruction* cond_instr = loop->GetConditionInst();
+  if (!cond_instr) {
+    return nullptr;
+  }
+
+  size_t iteration_count = 0;
+
+  // We have to check the instruction type here. If the condition instruction
+  // isn't of one of the below types we can't calculate the trip count.
+  switch (cond_instr->opcode()) {
+    case SpvOpULessThan:
+    case SpvOpSLessThan:
+    case SpvOpULessThanEqual:
+    case SpvOpSLessThanEqual:
+    case SpvOpUGreaterThan:
+    case SpvOpSGreaterThan:
+    case SpvOpUGreaterThanEqual:
+    case SpvOpSGreaterThanEqual:
+      if (loop->FindNumberOfIterations(
```

Comment:
We require these checks because FindNumerOfIterations only handles the listed opcodes. In debug it will assert if it is given an unsupported opcode, but may crash in release.

Reply:
I've cleaned the method a little by using IsSupportedCondition rather than the switch statement.